File: urbanairship-core/src/main/java/com/urbanairship/json/matchers/ExactValueMatcher.java
/* Copyright Airship and Contributors */

package com.urbanairship.json.matchers;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;

import com.urbanairship.json.JsonList;
import com.urbanairship.json.JsonMap;
import com.urbanairship.json.JsonValue;
import com.urbanairship.json.ValueMatcher;

import java.util.Map;

/**
 * Exact value matcher.
 *
 * @hide
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public class ExactValueMatcher extends ValueMatcher {

    @NonNull
    public static final String EQUALS_VALUE_KEY = "equals";

    private final JsonValue expected;

    /**
     * Default constructor.
     *
     * @param expected The expected value.
     */
    public ExactValueMatcher(@NonNull JsonValue expected) {
        this.expected = expected;
    }

    @NonNull
    @Override
    public JsonValue toJsonValue() {
        return JsonMap.newBuilder()
                      .putOpt(EQUALS_VALUE_KEY, expected)
                      .build()
                      .toJsonValue();
    }

    @Override
    protected boolean apply(@NonNull JsonValue value, boolean ignoreCase) {
        return isEquals(expected, value, ignoreCase);
    }

    public boolean isEquals(@Nullable JsonValue valueOne, @Nullable JsonValue valueTwo, boolean ignoreCase) {
        valueOne = valueOne == null ? JsonValue.NULL : valueOne;
        valueTwo = valueTwo == null ? JsonValue.NULL : valueTwo;

        if (!ignoreCase) {
            return valueOne.equals(valueTwo);
        }

        if (valueOne.isString()) {
            if (!valueTwo.isString()) {
                return false;
            }

            return valueOne.optString().equalsIgnoreCase(valueTwo.getString());
        }

        if (valueOne.isJsonList()) {
            if (!valueTwo.isJsonList()) {
                return false;
            }

            JsonList listOne = valueOne.optList();
            JsonList listTwo = valueTwo.optList();

            if (listOne.size() != listTwo.size()) {
                return false;
            }

            // iterate over both lists
            for (int i = 0; i < listOne.size(); i++) {
                if (!isEquals(listOne.get(i), listTwo.get(i), ignoreCase)) {
                    return false;
                }
            }

            return true;
        }

        if (valueOne.isJsonMap()) {
            if (!valueTwo.isJsonMap()) {
                return false;
            }

            JsonMap mapOne = valueOne.optMap();
            JsonMap mapTwo = valueTwo.optMap();

            if (mapOne.size() != mapTwo.size()) {
                return false;
            }

            // iterate over both maps
            for (Map.Entry<String, JsonValue> entry : mapOne) {
                if (!mapTwo.containsKey(entry.getKey())) {
                    return false;
                }

                if (!isEquals(mapTwo.get(entry.getKey()), entry.getValue(), ignoreCase)) {
                    return false;
                }
            }

            return true;
        }

        return valueOne.equals(valueTwo);
    }

    @Override
    public boolean equals(@Nullable Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        ExactValueMatcher that = (ExactValueMatcher) o;

        return expected.equals(that.expected);
    }

    @Override
    public int hashCode() {
        return expected.hashCode();
    }

}

File: urbanairship-core/src/main/java/com/urbanairship/json/matchers/ExactValueMatcher.kt
/* Copyright Airship and Contributors */
package com.urbanairship.json.matchers

import com.urbanairship.json.JsonValue
import com.urbanairship.json.ValueMatcher
import com.urbanairship.json.jsonMapOf

/**
 * Exact value matcher.
 *
 * @hide
 */
internal class ExactValueMatcher(
    private val expected: JsonValue
) : ValueMatcher() {

    override fun toJsonValue(): JsonValue  = jsonMapOf(
        EQUALS_VALUE_KEY to expected
    ).toJsonValue()

    override fun apply(value: JsonValue, ignoreCase: Boolean): Boolean {
        return if (ignoreCase) {
            equalsIgnoreCase(value, expected)
        } else {
            value == expected
        }
    }

    private fun equalsIgnoreCase(
        valueOne: JsonValue,
        valueTwo: JsonValue,
    ): Boolean {
        if (valueOne.isString  && valueTwo.isString) {
            return valueOne.optString().equals(valueTwo.optString(), ignoreCase = true)
        }

        if (valueOne.isJsonList && valueTwo.isJsonList) {
            val arrayOne = valueOne.optList()
            val arrayTwo = valueTwo.optList()

            if (arrayOne.size() != arrayTwo.size()) {
                return false
            }

            for (i in 0..<arrayOne.size()) {
                if (!equalsIgnoreCase(arrayOne[i], arrayTwo[i])) {
                    return false
                }
            }

            return true
        }

        if (valueOne.isJsonMap && valueTwo.isJsonMap) {
            val mapOne = valueOne.optMap()
            val mapTwo = valueTwo.optMap()

            if (mapOne.size() != mapTwo.size()) {
                return false
            }

            for ((key, childOne) in mapOne) {
                val childTwo = mapTwo[key] ?: return false
                if (!equalsIgnoreCase(childOne, childTwo)) {
                    return false
                }
            }

            return true
        }

        // Remaining types - bool, number
        return valueOne == valueTwo
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }

        val that = other as ExactValueMatcher

        return expected == that.expected
    }

    override fun hashCode(): Int {
        return expected.hashCode()
    }

    companion object {
        const val EQUALS_VALUE_KEY: String = "equals"
    }
}
